### PR TITLE
Adding support for APNS FB service.

### PR DESCRIPTION
**Pull request by chromalley**

### CODE REVIEW

File: src/main/scala/co/adhoclabs/task/Main.scala [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 
 import akka.actor.ActorSystem
+import co.adhoclabs.task.consumer.ApnsFeedbackConsumerActor
 import co.adhoclabs.task.demo.{ DemoSenderActor, DemoReceiverActor }
 
 /**
  * Created by yeghishe on 3/3/15.
  */
-object Main extends App {
+object Main extends App with Config {
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/Main.scala]: Revert this file.

File: src/main/scala/co/adhoclabs/task/TaskActor.scala [hunk 2]
```diff
@@ -20,11 +20,19 @@ class TaskActor extends Actor with ActorLogging with Config {
   protected def createProducerActor(taskType: TaskType): ActorRef =
     context.actorOf(CommonProducerActor.props(taskType), s"producer-${taskType.name}")
 
-  protected def createProxyConsumerActor(taskType: TaskType, targetActorPath: String): ActorRef =
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/TaskActor.scala]: Keep this code unchanged and move ApnsFeedbackConsumerActor actor creation to
https://github.com/adhoclabs/task/blob/master/src/main/scala/co/adhoclabs/task/TaskActor.scala#L47

File: src/main/scala/co/adhoclabs/task/TaskActor.scala [hunk 3]
```diff
@@ -20,11 +20,19 @@ class TaskActor extends Actor with ActorLogging with Config {
...
+  protected def createProxyConsumerActor(taskType: TaskType, targetActorPath: String): ActorRef = {
+
+    val path = ActorPath.fromString(targetActorPath)
+    val props = taskType match {
+      case TaskType.inactivePushToken ⇒ ApnsFeedbackConsumerActor.props(Option(apnsCertificatePath), Option(apnsCertificatePwd), path)
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/TaskActor.scala]: It would be much cleaner if instead of passing apnsCertificatePath, apnsCertificatePwd params from task actor to props then to actor you ApnsFeedbackConsumerActor actor gets it from the config. ApnsFeedbackConsumerActor is really the owner for those config items.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    apnsCertificatePwd: Option[String],
+    targetActorPath: ActorPath): Props =
+    Props(new ApnsFeedbackConsumerActor(apnsCertificatePath, apnsCertificatePwd, targetActorPath))
+}
+
+class ApnsFeedbackConsumerActor(apnsCertificatePath: Option[String],
+    apnsCertificatePwd: Option[String],
+    targetActorPath: ActorPath) extends CommonConsumerActor(TaskType.inactivePushToken, targetActorPath) with CamelApnsSupport {
+
+  override protected val certificatePath: Option[String] = apnsCertificatePath
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: I would also look into why this is optional, if it's not in config it should actually fail right there.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    new RouteBuilder() {
+      override def configure(): Unit = {
+        from("apns:consumer?initialDelay=10&delay=3600&timeUnit=SECONDS")
+          .to(path)
+      }
+    }
+  }
+
+  def handleInactiveDevices: Receive = {
+    case msg: CamelMessage if Try(msg.getBodyAs[InactiveDevice](classOf[InactiveDevice], camelContext)).isSuccess ⇒
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: I thinks failure case should be handled warned also. Anyway it makes sense to break this line.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala [hunk 6]
```diff
@@ -14,7 +15,8 @@ object CommonConsumerActor {
 }
 
 class CommonConsumerActor(override val taskType: TaskType, targetActorPath: ActorPath) extends BaseConsumer {
-  private case class Reply(msg: Any)
+  case class ConsumedMessage(msg: Any)
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: Can't find where this is being used.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala [hunk 7]
```diff
@@ -14,7 +15,8 @@ object CommonConsumerActor {
...
+  protected case class Reply(msg: Any)
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: Why was this changed to protected?

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala (hunk 6 above)
Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: It's not - will remove.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala [hunk 8]
```diff
@@ -24,18 +26,21 @@ class CommonConsumerActor(override val taskType: TaskType, targetActorPath: Acto
     super.preStart()
   }
 
+  protected def sendToTargetActor(msg: Any) = {
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: What's the return type of this method? Be explicit since it's protected.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala (hunk 5 above)
Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: So you think that this case to overwrite the one in the base class?

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala [hunk 9]
```diff
@@ -4,6 +4,7 @@ import akka.actor.{ ActorPath, ActorLogging, Props, Status }
 import akka.camel.{ Ack, CamelMessage, Consumer }
 import akka.util.Timeout
 import co.adhoclabs.task._
+import akka.pattern.{ ask, pipe }
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: Moving it back to L30 is good practice.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala (hunk 4 above)
Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: I think optionality can be removed entirely if we introduce protected methods that can be overwritten for tests.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala (hunk 7 above)
Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: Reply is used in sendToTargetActor, so it can no longer be private.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala (hunk 8 above)
Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: It's Unit. I'll add it.

File: src/test/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActorTest.scala [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+class ApnsFeedbackConsumerActorTest extends TestKit(ActorSystem("test-system"))
+    with Config
+    with ImplicitSender
+    with WordSpecLike
+    with Matchers
+    with BeforeAndAfterAll {
+
+  "ApnsFeedbackConsumerActorTest" should {
+
+    "send InactivePushToken message to target actor and pipe reply to self and send Ack to sender" in withConsumerActor { (consumerActor, targetProbe) ⇒
```
Comment (yeghishe) [inline src/test/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActorTest.scala]: Don't need to be that specific about pipe and reply, not that test can even test it, all it cares is that target actor gets it.

File: src/test/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActorTest.scala [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      EventFilter.warning(message = s"Got unexpected message: $msg", occurrences = 1) intercept {
+        consumerActor ! msg
+      }
+    }
+  }
+
+  private def withConsumerActor(testCode: (ActorRef, TestProbe) => Any): Unit = {
+    val targetProbe = TestProbe()
+
+    class TestConsumerActor extends ApnsFeedbackConsumerActor(None, None, targetProbe.ref.path) {
```
Comment (yeghishe) [inline src/test/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActorTest.scala]: Ok now I see how this is configured and why those are optional. I thinks it's a way to prevent test from using the certs.
configureApnsService flat maps it so a better way would be to have test override that method to void.

**Comment by yeghishe** [comment]
To give more context about apnsCertificatePath and apnsCertificatePass not being option[string]. it should be converted to string but still be optional. user should not be forced to specify those unless in config they specify that they're starting that consumer. if consumer is specified but options aren't there in config throw and exception with the right message so the app doesn't event start.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala (hunk 4 above)
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: Not sure I get it.

File: src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala (hunk 5 above)
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/ApnsFeedbackConsumerActor.scala]: no just have some thing like
case msg: CamelMessage => // short line here
  Try
  if success -> do what you need to do
  else -> write a warning

It will make it cleaner and logs won't yell that there was unhanded message.

File: src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala (hunk 7 above)
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: I feel like it breaks the encapsulation. A better way would be to have a case class in base that's protected, called SendToTarged(msg), base handles that method and both base and dependent can send that message to self.

Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/consumer/CommonConsumerActor.scala]: I'll give it another shot. I tried that with the ConsumedMessage case class, but I was getting errors when integrating it in with biz service.

File: src/main/scala/co/adhoclabs/task/CamelApnsSupport.scala [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    apnsServiceFactory.setCertificatePassword(apnsCertificatePwd)
+    val apnsService = apnsServiceFactory.getApnsService
+    val apnsComponent = camelContext.getComponent("apns", classOf[ApnsComponent])
+    apnsComponent.setApnsService(apnsService)
+  }
+
+  protected def addFeedbackRoute(camelContext: DefaultCamelContext, path: String) =
+    camelContext.addRoutes(new RouteBuilder() {
+      override def configure(): Unit = {
+        from("apns:consumer?initialDelay=10&delay=3600&timeUnit=SECONDS")
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/CamelApnsSupport.scala]: Do we want to move those to config file? Also I see those are the defaults in camel.

Comment (chromalley) [inline src/main/scala/co/adhoclabs/task/CamelApnsSupport.scala]: I didn't realize those were defaults. If so, we might as well just remove them altogether then, no?

File: src/main/scala/co/adhoclabs/task/TaskActor.scala [hunk 13]
```diff
@@ -41,8 +41,14 @@ class TaskActor extends Actor with ActorLogging with Config {
 
     log.debug(s"Starting consumers: $consumers, $consumerProxyActors")
 
-    consumers.map { t ⇒
-      consumerProxyActors.get(t).fold(createConsumerActor(t))(createProxyConsumerActor(t, _))
+    consumers.map { task ⇒
+      consumerProxyActors.get(task).fold(createConsumerActor(task)) { p ⇒
+        val path = ActorPath.fromString(p)
+        createProxyConsumerActor(task match {
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/TaskActor.scala]:
```
val props = if (task == TaskType.inactivePushToke) ApnsFeedbackConsumerActor.props(path) else CommonConsumerActor.props(task, path)
createProxyConsumerActor(props, task)
```
would be cleaner

File: src/main/scala/co/adhoclabs/task/package.scala [hunk 14]
```diff
@@ -22,7 +23,8 @@ package object task {
       outboundSms.name -> outboundSms,
       slackMessage.name -> slackMessage,
       segment.name -> segment,
-      push.name -> push
+      push.name -> push,
+      inactivePushToken.name → inactivePushToken
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/package.scala]: Let's change the rest to one character -> in this map also, to be consistent... also it looks better ))

File: src/main/scala/co/adhoclabs/task/worker/PushActor.scala [hunk 15]
```diff
@@ -13,41 +14,19 @@ import scalaz.{ -\/, \/- }
...
-    val apnsServiceFactory = new ApnsServiceFactory(camelContext)
-    apnsCertificatePath.foreach(apnsServiceFactory.setCertificatePath)
-    apnsCertificatePwd.foreach(apnsServiceFactory.setCertificatePassword)
-    val apnsService = apnsServiceFactory.getApnsService
-    val apnsComponent = camelContext.getComponent("apns", classOf[ApnsComponent])
-    apnsComponent.setApnsService(apnsService)
-  }
-
   @throws[Exception](classOf[Exception])
   override def preStart(): Unit = {
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/worker/PushActor.scala]: No {} needed 
```
override def preStart(): Unit = configureApnsService(camelContext)
```

**Comment by yeghishe** [comment]
At which point are you checking if config is correct for ApnsFeedbackConsumerActor? Meaning
when user puts
`consumers= ["inactive-push-token"]`
and no cert config it should throw config error.

**Comment by yeghishe** [comment]
Also please add a section in for apns feedback... How consumer is specified, how certs should be configured, if inactive-push-token is specified what the values should be, etc...

File: src/main/scala/co/adhoclabs/task/Config.scala [hunk 16]
```diff
@@ -27,9 +27,22 @@ private[task] trait Config {
...
-  private val pushConfig = taskConfig.getConfig("push")
-  val apnsCertificatePath = pushConfig.getString("certificate-path")
-  val apnsCertificatePwd = pushConfig.getString("certificate-password")
+  val pushConfig = taskConfig.getConfig("push")
+
+  val (apnsCertificatePath, apnsCertificatePwd) = if (consumers.contains(TaskType.inactivePushToken)) {
+    val path = pushConfig.getString("certificate-path")
+    val pass = pushConfig.getString("certificate-password")
+    if (path.isEmpty) {
+      throw new RuntimeException("certificate-path is missing")
```
Comment (yeghishe) [inline src/main/scala/co/adhoclabs/task/Config.scala]: I thought you wanted to use Missed exception
